### DISCUSSION TITLE
Dynamic routing: how to use default route "/"

### DISCUSSION THREAD

**Discussion by reposman33** [vercel/next.js#24072] [Help]
Hello,

I work through the Nextjs Learn tutorials. My question concerns the dynamic routing.
With non-dynamic routes  all React components in the /page folder map to the url slug. For example Component `/pages/about.js` maps to  ```localhost:3000/about```. Except for the root (`/`) default path: ```localhost:3000``` maps to `index.js`. Great, so we have a landing page here.

Now with dynamic routing the situation is alike: I have a directory with markdown files that are being dynamically mapped to the url route when using a dynamic file (`/pages/[page].js`). In `getStaticPaths` I read and return all filenames in a folder (for example in `/markdown`) and when a url route is requested `getStaticProps` render the corresponding  markdown file. So for example ```localhost:3000/projects``` renders `/markdown/projects.md`

But now my question: how can I render a root (or default) route like ```localhost:3000``` (or ```localhost:3000/```)? I can't create a file named `/md`.... I get a 404 when requesting ```localhost:3000```. A solution is to use a custom `404.md` as the default route but I don't think that's what it is for...

Am I missing something here or is dynamic rendering not suitable in this case?

**Comment by akellbl4** [comment]
If I understood right. You need to create pages like that:
```
- pages
  - index.js
  - 404.js
  - [page].js
```
As I know named files have higher priority when Next.js maps routes. It means when you create a named route it has priority on dynamic mapping.

**Reply by reposman33** [reply]
Yes, that is true. I guess my problem lies elsewhere.

When I use dynamic routes in [path].js I use getStaticPaths to get a listing of all files in /markdown. I use this listing to generate a menu. As long as I use the dynamic route [page].js this works - in this component I retrieve the listing using getStaticPaths and node.fs and inject those in a React component. That component displays a menu based upon the filenames.
For the index.js file I also want this. But in index.js I cannot import node.fs and use getStaticPaths _or_ import the result of getStaticPaths from [page].js without getting the error 

```
./lib/pages.js:1:0
Module not found: Can't resolve 'fs'
> 1 | import fs from "fs";
```
Apparently the node.fs dependency is not tree shaken and sent to the client.  I found a trick to create an empty fs component in next.config.js

```
module.exports = {
	future: {
		webpack: (config, { isServer }) => {
			// Fixes npm packages that depend on `fs` module
			if (!isServer) {
				config.node = {
					fs: "empty",
				};
			}
			return config;
		},
	},
};
```

but that doesn't prevent the error.

How can I do this? Do I have to use getServerSideProps in the index.js file instead?

**Reply by akellbl4** [reply]
Hm, got it.
As I can see you put `webpack` inside `future` but it should be like:
```js
module.exports = {
    webpack(config) {
      return config;
    },
};
```

**About fetching:**
`getServerSideProps` enables SSR and I think it's not what you wanted.
`getStaticProps` should work just fine in your situation.

**Null dependencies**
If I need to null module for frontend and it wasn't treeshaked. I use `null-loader`

```js
module.exports = {
  webpack: (config, { isServer }) => {
    // Fixes npm packages that depend on `fs` module
    if (!isServer) {
      config.module.rules.unshift({
        test: /^fs$/,
        use:'null-loader'
      })
    }
    return config;
};
```

**Comment by reposman33** [comment]
Thanks, appreciate the help.

Still it can't achieve what I want:

I have a dynamic route [page].js that has the getStaticPaths and
getStaticProps methods, GetStaticPaths retrieves all filenames from a
folder containing md files. That data is via props injected in a Navigation
Component. That component uses the array to create a menu - Links that
allow you to view the routes. This works fine and I can switch routes by
clicking those Links.

But I can't implement the default route (/) this way.
I _can_ simply add pages/index.js and then when I navigate to
localhost:3000 that component is rendered. But it doesn't render the menu -
for this I have to rertieve all filenames like I did in [page].js. and that
seems to be impossible because I need the fs library and that is not
available client-side. I keep getting the error message

./lib/pages.js:1:0Module not found: Can't resolve 'fs'> 1 | import fs from "fs";

even when I export only the array of filenames from ../lib/pages.js. The
whole component is loaded along with the fs dependency .


Is there a way to achieve this?

**Reply by akellbl4** [reply]
Could you please provide GitHub repo with an example of the error?

**Comment by reposman33** [comment]
Hi Pavel,

I decided to start over the Next.js tutorial. I try to do too many things
at the same time.

Thanks for your help, I might come back later
